### DISCUSSION TITLE
How do AOVs work?

### DISCUSSION THREAD

**Discussion by yusrig** [bnpr/Malt#583] [Q&A] [answered]
Hello. I am trying to get Custom AOV outputs to work. My understanding so far is as follows:

When I create a custom output in my Mainpass Output node I can pass in any AOV type I want from the mesh nodes. But then I need to pass this into the Render Layer as a new output so that it can be accessed in the Render nodetree and then in turn be passed into the Screenpass nodetree. I think I have wired this up correctly (correct me if I am wrong) and should be getting custom AOVs in my render. But I am not. I attached screenshots of what I did below. Can someone help me figure out why this isn't working as I expect it to?

P.S. I am not an advanced user and some of the detailed concepts are a little confusing for me. If you could add context would be great!

1. First I need to create a custom output in my mesh graph:
![image](https://github.com/user-attachments/assets/e00eeaf3-b302-4e75-85bc-54bc8fbb37f8)
2. Then I need to make sure in my render layer I connect that to render layer output so I can then access it in the render graph
![image](https://github.com/user-attachments/assets/0346eac2-7703-4e29-898b-7e0b5ca2c7c7)
3. In the render graph I pass it into the screen pass which I have added custom input and output for this custom AOV.
![image](https://github.com/user-attachments/assets/c50a9cc6-fb3c-4b9f-8169-ad4fa78c34b1)
4. Here is what my screen pass looks like:
![image](https://github.com/user-attachments/assets/25b5b5b9-7421-4457-979e-985e5caa0af7)
5. But when I render i only get a black image with no data under the custom output
![image](https://github.com/user-attachments/assets/248a97e9-06a8-4ba3-8484-9b3cf5867d9a)

**Comment by Warrenayiss** [comment, marked as answer]
I'm not skilled in this, in fact, I used your explanation to reproduce the bloom effect shared in #347 so I will tell you where what I did differ from what you explained.
In step 3, I didn't change the default ScreenPass I created a different ScreenPass node that I put between it and SuperSamplingAA

![image](https://github.com/user-attachments/assets/3a010398-e7f6-4d19-bf82-2ab3900bfd0e)

Then made a completely new Screen node tree
![image](https://github.com/user-attachments/assets/3c41f78d-b8cc-4ae2-931c-ea651bb9dc8e)

Hope this help you at least a little bit
And thank you teaching me about AOVs

**Reply by yusrig** [reply]
Oh man, am I dumb! Yes, your solution helped, creating a new Screen pass node tree was required to get the AOV to work, instead of using the same Screenpass. But more over I made a critical mistake in my Mainpass output. I used a Vec3 instead of Vec4 for the custom AOV output and thats why it wasn't rendering anything (I think...).

Anyway using a separate screenpass helped. So I guess I could make multiple Screen pass materials to be different AOVs and get different passes for compositing later.

**Reply by pragma37** [reply]
You should be able to connect `RenderLayers.shadow` directly to `RenderOutput.shadow`.
I think the only reason your initial setup didn't work was because of the `vec3` output (If each layer output has no transparency then they can't be composed together, so you only see the result of the last layer, which is probably empty).
